### DISCUSSION TITLE
Bug with Python MySQLdb library and 20s+ db connection (not transaction).

### DISCUSSION THREAD

**Discussion by mattwalshdev** [planetscale/discussion#557] [Bug Reports] [answered]
Bug replicated on Python 3.9, 3.10, 3.11

Finally worked this one out, been getting the below every now and then
_target: primarydb.-.primary: vttablet: rpc error: code = Aborted desc = transaction <transaction>: ended at <time> (exceeded timeout: 20s)_

I have been committing after every transaction, and looking at the PlanetScale logs the longest transaction is around 2 seconds.

The issue is with the MySQLdb Python library when used with PlanetScale
https://mysqlclient.readthedocs.io/ 
https://pypi.org/project/mysqlclient/

If you create a new connection to PlanetScale, and then run a loop of any mysql operations even if you are committing after each transaction, you will see the above error after 20 seconds, when I believe the longest running connections allowed are 900 seconds?

**Working solution: ** I have installed an alternate Python mysql library, pymysql https://pypi.org/project/pymysql/ and it works without the error on 20+ second routines.

I was in the process of migrating to pymysql anyway as it's required for my isolated AWS lambdas, as the MySQLdb library is linked to a MySQL engine you must install on your linux server, and lambdas don't allow you to install additional linux packages on the server easily and would slow down the lambda greatly waiting for install everytime.

As there is a working library without additional linux package requirements, the solution may just be to make note somewhere in docs to not use MySQLdb Python library with PlanetScale.

**Comment by orware** [comment, marked as answer]
Hello @mattwalshdev,

Thank you for sharing your experience and feedback here when using transactions extensively as this could be a good tip for others as well!

A few related questions: for either library, are you making use of the `autocommit = True` flag at all? With Python connectors defaulting to `False` this can also result in an increased likelihood of users encountering transaction timeouts, since the default behavior will implicitly make even `SELECT` queries kick off the 20 second transaction timer unintentionally.

If `autocommit = True` is already in use and you were still encountering that issue even with regularly committing throughout the process with the regular Python `mysqlclient` library, and `PyMySQL` didn't experience the same issue then that definitely seems like an important (and helpful) behavior difference for users to be aware of.

---

Our Developer Education team also put together the following Python related post last year covering basic connections to several different Python MySQL libraries to share as an additional reference for anyone else coming to this discussion in the future:
https://planetscale.com/blog/connect-to-a-mysql-database-in-python

**Comment by mattwalshdev** [comment]
Thanks orware, tested and that does solve it for the MySQLdb library

```python
import MySQLdb
from MySQLdb.connections import Connection
from MySQLdb.cursors import Cursor, DictCursor
import time


#ssl location may differ on linux builds
con: Connection = MySQLdb.connect(
    host="planetscale_host",
    user="planetscale_user",
    passwd="planetscale_password",
    db="planetscale_db",
    ssl_mode="VERIFY_IDENTITY",
    ssl={"ca": "/etc/ssl/certs/ca-certificates.crt"},
)
#cur: DictCursor = con.cursor(cursorclass=DictCursor)
cur: Cursor = con.cursor()

start_time = time.time()
while True:
    try:
        cur.execute(f'SELECT * FROM `A_TABLE`', ())
        results = cur.fetchall()
        print(f"Running for {time.time() - start_time} seconds")
        time.sleep(2)
    except Exception as e:
        print(e)
        print(f"Program crashed in {time.time() - start_time} seconds")
        exit()
```

Without autommit the above returns -
```
Running for 0.05556464195251465 seconds
Running for 2.1146621704101562 seconds
Running for 4.173159122467041 seconds
Running for 6.229693174362183 seconds
Running for 8.291164636611938 seconds
Running for 10.349453210830688 seconds
Running for 12.40724492073059 seconds
Running for 14.466468095779419 seconds
Running for 16.52493453025818 seconds
Running for 18.583406448364258 seconds
Running for 20.641520977020264 seconds
(1105, 'target: xxx.-.primary: vttablet: rpc error: code = Aborted desc = transaction xxx: ended at 2023-10-31 00:57:25.884 UTC (exceeded timeout: 20s) (CallerID: xxx)')
Program crashed in 22.698744297027588 seconds
```
but adding `autocommit=True` to the connection arguments resolves. It might be worth adding that in as an argument in the article you linked as it was tricky to diagnose and this issue has been reported before https://github.com/planetscale/discussion/discussions/219

Edit - just realised that con.commit() after each statement resolves too 🤦 I haven't been committing after SELECT as I didn't realise any transaction to store took place after a SELECT.
